Fixed #8936 -- Added view permissions and a read-only admin [rebased]

olivierdalang opened the pull request:
Hi,

This is just a rebase of https://github.com/django/django/pull/5297
The only difference (besides fixing the merge conflicts) is that `ModelAdmin.declared_fieldsets` doesn't exist anymore and was replaced by `get_fieldsets`.

Thanks,

Olivier

auvipy commented:
@olivierdalang it would be great if you could fix the flake errors.

olivierdalang commented:
@auvipy Ok I think flake errors are fixed (hehe didn't know about this tool)

auvipy commented on docs/ref/contrib/admin/actions.txt:
```diff
@@ -34,6 +34,16 @@ models. For example, here's the user module from Django's built-in
     deletion <topics-db-queries-delete>`.
 
+.. note::
+
```
this should be 1.11 as 1.10 is feature freezed

auvipy commented on docs/ref/contrib/admin/index.txt:
```diff
@@ -1722,6 +1722,18 @@ templates used by the :class:`ModelAdmin` views:
 
+.. method:: ModelAdmin.has_view_permission(request, obj=None)
+
+    .. versionadded:: 1.10
```
1.11

auvipy commented on docs/ref/models/options.txt:
```diff
@@ -331,12 +331,16 @@ The name of the manager to use for the model's
     the model is created by :djadmin:`migrate` in order to prevent any omitted
     permissions from being created.
 
+    .. versionchanged:: 1.10
```
1.11

auvipy commented on docs/releases/1.10.txt:
```diff
@@ -72,6 +72,22 @@ Minor features
   link <django.contrib.admin.AdminSite.site_url>` at the top of each admin page
   will now point to ``request.META['SCRIPT_NAME']`` if set, instead of ``/``.
 
```
you need to move the changes of this part to 1.11.txt

olivierdalang commented:
@auvipy I made the changes, thanks for the detailed review ! (and sorry for the mistakes, it's my first contribution to django)

auvipy commented:
No worries!! you did some very well!!!

timgraham commented:
buildbot, test this please.

timgraham commented:
The test failures are:
admin_views.tests.UserAdminTest.test_user_fk_change_popup
admin_views.tests.UserAdminTest.test_user_fk_delete_popup

timgraham commented:
Did you fix https://github.com/django/django/pull/5297#issuecomment-234072204?

olivierdalang commented:
Hi,

It seems these tests fail because of a newline in the template (to comply with flake) :
```
self.assertContains(response, 'class="related-widget-wrapper-link change-related" id="change_id_owner"')
```
```
AssertionError: False is not true : Couldn't find 'class="related-widget-wrapper-link change-related" id="change_id_owner"' in response
```

while this appears in the template :

```
        <a class="related-widget-wrapper-link change-related"
            id="change_id_owner"
```

An easy fix is to add the line break in the test, but this will fail again if the template gets changed. I see there's the `html` parameter to `assertContains`, but it doesn't seem to work with partial match (the test only tests for the `class` and the `id` attribute, it doesn't match since there are also a `data-href-template` and a `title` html attributes.

timgraham commented on django/contrib/admin/templates/admin/related_widget_wrapper.html:
```diff
@@ -2,11 +2,17 @@
     {% block links %}
-        {% if can_change_related %}
-        <a class="related-widget-wrapper-link change-related" id="change_id_{{ name }}"
+        {% if can_change_related or can_view_related%}
```
missing space before "%}"

timgraham commented:
I don't have a strong opinion about how to fix that.

olivierdalang commented:
Hi,

I made the changes :
- missing space & flake8
- failing tests (I just added the line break in the tested string, it's not very elegant but I don't see how to do it without making the test case too complex)
- https://github.com/django/django/pull/5297#issuecomment-234072204 including another edge case when the inline has view and delete permissions

I intend to squash those latter commits if they look good to you.

I suggest closing #5297 to avoid having comments at different places.

I removed the patch needs improvement flag on the ticket. Let me know if I can do anything more for this !

olivierdalang commented:
There's a failing test, but it looks like it's travis fault : `FATAL: Could not checkout ea3c082f7160685ab0c8215745dd191e0a85424c` 
Any way to restart the tests ?

timgraham commented:
Rebase so the patch merges cleanly and squash commits (in general, there's no need to push incremental commits for minor changes; just squash right away). Of course with a patch of this size, after someday reviews the patch the first time , it might be useful to push incremental changes so they don't have to recheck the entire patch again.

yigitguler commented:
This will be a nice addition to the Django Admin. Can't wait to see it!

olivierdalang commented:
Ok I think the PR is ready !

slurms commented:
We've been testing this in our own admin site and found a couple of things:
 - using `prepopulated_fields` with view only permissions fails (tries to access a field which doesn't exist on the form due to excludes).
 - using custom admin forms with form fields that only exist on the form fails (`label_for_field` issue).

I'll try and write a couple of failing test cases.

hirokiky commented:
Hi. It's awesome PR and I tried it with my existing web app and existing database..
My question is how can I migrate from existing database? 
As the Options.default_permissions doc said [here](https://github.com/django/django/pull/6734/files#diff-441cf273fa22534a25e767bb35062b7aR337), old existing apps can't use this new feature. 

I tried recreating the database and tried this feature, It worked with my app.

pope1ni commented:
@hirokiky: I see no reason why old existing apps cannot use this feature - the note states that backwards compatibility is maintained by ensuring that having a _change_ permission also implicitly means that the user has the _view_ permission for the same model. This is to avoid loss of access to pages in the admin site when Django is upgraded because users would not have the new _view_ permissions by default. After upgrade _change_ permissions can be removed and _view_ permissions added as required. (It also rarely makes sense to be able to change something but not view it!)

hirokiky commented:
@pope1ni Yes. Admin pages can be seen for existing users.

But with migrated DB, I can't apply `Can view <model>` for users. there isn't any `Can view <model>` permissions at change users page on admin.
The reason is that there aren't any `Can view <model>` permission on Permission model.

pope1ni commented on tests/admin_views/admin.py:
```diff
@@ -150,6 +150,10 @@ def has_change_permission(self, request, obj=None):
 
+    def has_view_permission(self, request, obj=None):
+        """ Only allow viewing objects with even id number """
+        return request.user.is_staff and (obj is not None) and (obj.id % 2 == 0)
```
I haven't run the tests, so I could be off here, but if having the _change_ permission effectively grants you the _view_ permission for backwards compatibility, is having the same criteria here a good idea? It could potentially mask problems with the new _view_ permission in the tests as the user would always have both.

pope1ni commented on django/contrib/contenttypes/admin.py:
```diff
@@ -98,7 +98,7 @@ def get_formset(self, request, obj=None, **kwargs):
         else:
             exclude = list(self.exclude)
-        exclude.extend(self.get_readonly_fields(request, obj))
+        exclude.extend(self.get_uneditable_fields(request, obj))
```
Where is `get_uneditable_fields()` defined?

pope1ni replied:
Ah. Ignore this - GitHub had hidden the diff of `django/contrib/admin/options.py` when I skimmed through last time.

pope1ni commented on docs/ref/contrib/admin/index.txt:
```diff
@@ -1746,6 +1746,18 @@ templates used by the :class:`ModelAdmin` views:
+    will be interpreted as meaning that the current user is not permitted to
+    view any object of this type).
+    The ``view`` permission has been made with backward compatibility in mind, so
+    ``change`` permission imply that user can also view objects.
```
Although the change is backwards compatible by allowing users who can _change_ to implicitly be allowed to _view_, and this makes sense because you have to see something to edit it, there is still a problem. Existing models will not have a `Permission` object for the _view_ permission.

There needs to be some documentation explaining how to add the missing permissions or perhaps a management command to create permissions objects for all models where they are found to be missing. I'll defer to @timgraham on this.

bittner commented:
Any chance this feature will end up in Django 1.11? Not too much time left, right?

auvipy commented:
1.11 is already feature freez

timgraham commented:
New features aren't allowed after the alpha release, see https://code.djangoproject.com/wiki/Version1.11Roadmap.

olivierdalang commented:
Hi all,

@pope1ni Thanks a lot for the detailed review. I made some of the changes (the easy picks).

I didn't do the initial implementation  Do you think there could be a better solution than get_uneditable_fields() ?

So here's the todo list to get this ready, am I missing something ?

- [ ] Fix problems reported by @slurms  (@slurms, can you provide the tests ?)
- [ ] Make it clear what to do to deal with old apps so that the view permission is created (@timgraham, any pointer on this ? I kind of naively thought the permissions would be created with migrations)
- [ ] Explain get_uneditable_fields in the code (@PetrDlouhy , since you did the initial implementation, could you help me on this ? thanks again by the way)
- [ ] Rebase on master

Just a disclaimer : this is my first serious contribution to django, so reviewers please don't assume I'm exactly knowing what I'm doing

Let's hope it will make it into Django one day :)

olivierdalang commented:
Ok I just rebased, it was less painful than expected.

- [ ] Fix problems reported by @slurms  (@slurms, can you provide the tests ?)
- [ ] Make it clear what to do to deal with old apps so that the view permission is created (@timgraham, any pointer on this ? I kind of naively thought the permissions would be created with migrations)
- [ ] Explain get_uneditable_fields in the code (@PetrDlouhy , since you did the initial implementation, could you help me on this ? thanks again by the way)
- [X] Rebase on master

olivierdalang commented on tests/admin_views/tests.py:
```diff
@@ -1704,6 +1753,18 @@ def test_change_view(self):
+            response = self.client.get(change_url_3)
+            self.assertEqual(response.status_code, 200)
+            response = self.client.post(change_url_3, {'name': 'changed'})
+            self.assertEqual(response.status_code, 302)
```
Is it normal that the error code is 302 redirect ? I was expecting a 403 as well.

wengole replied:
Yes as the standard response is to redirect to the login page

pope1ni commented on django/contrib/admin/helpers.py:
```diff
@@ -225,44 +225,53 @@ class InlineAdminFormSet:
     def __init__(self, inline, formset, fieldsets, prepopulated_fields=None,
-                 readonly_fields=None, model_admin=None):
+                 readonly_fields=None, uneditable_fields=None, has_add_permission=True,
+                 has_change_permission=True, has_delete_permission=True, model_admin=None):
```
Should we add `has_view_permission` for consistency here? I'm not sure I like that we assume that users have these permissions by default. Should these not be added after `model_admin` to not break backwards compatibility of parameter ordering? Currently only `has_add_permission` gets used in this class and all three of these flags are used at the same time outside of the class definition. Would it make more sense to pass in a `view_only=True` or `editable=False` flag if we do not need the granularity?

As previously mentioned, I think that the overlap between `readonly_fields` and `uneditable_fields` is messy. I don't know whether we can make this work with `readonly_fields` alone or whether we need to come up with another solution.

olivierdalang replied:
Ok I'll add had_view_permission as well and put the news parameters at the end. I think has_*_permissions are more explicit than flags.

pope1ni commented on django/contrib/admin/options.py:
```diff
@@ -1009,12 +1047,21 @@ def render_change_form(self, request, context, add=False, change=False, form_url
             'change': change,
             'has_add_permission': self.has_add_permission(request),
             'has_change_permission': self.has_change_permission(request, obj),
+            'has_view_permission': self.has_view_permission(request, obj),
```
Move `has_view_permission` above `has_add_permission` for consistency.

pope1ni commented on django/contrib/admin/options.py:
```diff
@@ -1953,7 +2032,7 @@ def get_fields(self, request, obj=None):
     def get_queryset(self, request):
         queryset = super().get_queryset(request)
-        if not self.has_change_permission(request):
+        if not (self.has_change_permission(request) or self.has_view_permission(request)):
```
`if not ... and not ...` for consistency with other similar statements...

olivierdalang commented:
@pope1ni thanks for the detailed review. I made all changes about style improvement and am starting to look into the three good questions you ask

olivierdalang commented on django/contrib/admin/options.py:
```diff
@@ -1902,43 +1960,64 @@ def get_formset(self, request, obj=None, **kwargs):
+            can_change = self.has_change_permission(request, obj)
+            can_add = self.has_add_permission(request)
+
+        defaults['form'] = get_model_form(can_add, can_change)
```
I must admin I'm not sure I completely understand what's happening here. As I said, I didn't do the initial implementation.

As far as I get it, despite it's name, the DeleteProtectedModelForm class is not only about deletion, but simply the base_model_form used for inlines. It seems to be needed to validate the DELETE field.

Then, about defaulting to true if there's no request, it's a good question. I'm not sure in which contexts it can happen that there's no request. I guess it's only outside of the regular admin context ? Or can it even happen ?

pope1ni commented on docs/ref/contrib/admin/actions.txt:
```diff
@@ -34,6 +34,12 @@ models. For example, here's the user module from Django's built-in
+
+    .. versionchanged:: 2.0
+
+    The new ``view`` permission has made it possible to access pages in the admin site without the ability to make changes. Actions can only be performed if a user has the ``change`` permission.
```
Wrap documentation to 80 characters.

pope1ni commented on django/contrib/admin/options.py:
```diff
@@ -455,6 +472,21 @@ def has_change_permission(self, request, obj=None):
         codename = get_permission_codename('change', opts)
         return request.user.has_perm("%s.%s" % (opts.app_label, codename))
 
+    def has_view_permission(self, request, obj=None):
```
I think you missed this one in your recent updates.

PetrDlouhy commented:
@olivergeorge Thank You for pushing this forward. The `uneditable_fields` are fields, that user can't edit either because he doesn't have permission or their edition is forbidden by admin settings. With introduction of view only permissions we need to be able to display all fields in read only mode.

pope1ni commented:
@PetrDlouhy Thank you for the initial work on this improvement. I have had time to think about this a little.

My main objection to the addition of `uneditable_fields` is that it seems exactly the same as `readonly_fields`, just reworded. Essentially if the user does not have permission to add or change (depending on the context) we really just want to dynamically append to `readonly_fields` for that request.

This could be achieved overloading `get_readonly_fields` with the permission checks although we would have to ensure that users who overload that themselves call `super()`. Alternatively there should be a private method/property that is used in place of `get_readonly_fields`, calls `get_readonly_fields` and then extends the result based on the permissions. This output can then be passed on down the chain to inline forms, etc.

My second problem is that the `uneditable_fields` argument / property and `get_uneditable_fields` method currently look as though they would be exposed as public API and this would confused users as the readonly_fields stuff exists already and this is really only intended to allow this new view permission to work.

My third issue is that `uneditable_fields` and `readonly_fields` are currently being passed around interchangeably which feels wrong and error-prone.

slurms commented on django/contrib/admin/options.py:
```diff
@@ -313,6 +314,22 @@ def get_readonly_fields(self, request, obj=None):
+        local_fields = [field.name for field in self.opts.local_fields]
+        local_many_to_many = [field.name for field in self.opts.local_many_to_many]
+        readonly_fields = list(self.get_readonly_fields(request, obj))
+        local_fields = [field.name for field in self.opts.local_fields]
```
I don't think this is needed (probably accidentally copied from line 327).

ryselis commented:
We have implemented a detail view in a different way.

We have added `detail_view` method to model admin and a different template. Properties `fields_detail` and `fieldsets_detail`and methods `get_fields_detail` and `get_fieldsets_detail` to define fields and fieldsets. View permission and `has_view_permission` have been added as well. Wouldn't this approach be better approach? We are willing to share the code.

pope1ni commented:
@ryselis: What you suggest seems like a lot of copy and paste to provide a second copy of code used soley for the view-only case which is not going to be acceptable. If I am wrong feel free to comment with a link to a branch with your proposition.

ryselis commented:
@pope1ni  It depends. We have used detail view for 2-3 years and at least for us it has diverged from the change view. Main differences:

- we have implemented tabs instead of inlines. They are not just read-only inlines placed in tabs. They work more like mini changelists. Assume this data model: inquiry is the base model, commercial offer has a FK to inquiry (one inquiry may have multiple commercial offers) and commercial offer has its own inline model commercial offer item. There is no way to put commercial offer as inline in inquiry admin because it needs its own inlines, but it very reasonable to have a list of related commercial offers in inquiry detail, and with custom fields as well, like total sum of all items.
- different fields are needed in change and detail forms, it is reasonable to show images and even charts there
- different buttons are needed in change and detail forms

Basically detail view is a blend of change form and changelist view for single item with additions - at least for how we are using it.

I have hacked a detail view from our project into django so that you can see how much code would need to be duplicated. This code is not tested with unit tests and stripped to a minimal example. If we decide that this approach is acceptable, I will write tests and make the code more beautiful.
 [Here is the branch with the code, forked from master](https://github.com/ryselis/django/tree/issue_8936_detail_view).
We use a form under the hood to allow editing some fields from detail view - this might not be the best idea, it could be changed to something else.

jfscyberpub commented:
Hello everyone,

Someone wrote this project: https://github.com/vint21h/django-read-only-admin
I don't know if it can help you but it might be worth sharing it.

Have a great day!

olivierdalang commented on django/contrib/admin/helpers.py:
```diff
@@ -225,44 +225,53 @@ class InlineAdminFormSet:
-            self.readonly_fields, model_admin=self.opts,
-        )
+
+        if self.has_add_permission:
```
There's no check for the change permission, because even without it, the inlines are shown (only they are shown with readonly fields)

viktor25 commented:
Hello everyone,
How far is this feature from being merged?
I'm hoping to see it in Django 2.0.  If someone could give me some direction, I'd be happy to help.

olivierdalang commented:
Hi,

From my point of view, this is ready to merge.

I think @slurms problems happens because he has setup some custom admin forms, so when the form is read only, he tries to access inexistant fields. I don't think this should be considered as a bug, but it's true that with read only admin, the read-only case must now be considered when using custom forms. I wrote some notes about that in `docs/release/2.0.txt`.

@viktor25 If you want to help make this into 2.0, please test the branch with your projects to see if you have issues and review the PR.

Also I will update the ticket to remove "patch needs improvement", so that it should be picked up by @timgraham (forgot about that last time :) ). Tim please let me know if there's anything more I can do to get this into 2.0.

pope1ni commented:
Hi @olivierdalang. I'll try to review this again sometime this week.

viktor25 commented:
I tested with my projects, here are the results.

Regressions:
1. JSONField with value [1, 2] is shown as 1, 2 instead of [1, 2].  JSONField with value [] is shown as empty.
2. Geographic locations do not display on a map.  (Model with a django.contrib.gis.db.models.PointField, admin model inherits from django.contrib.gis.admin.OSMGeoAdmin)
3. A ManyToManyField with lots of values displays as a wall of text.  It should display as a limited-size scrollable container.
4. A foreign key in raw_id_fields should be displayed as a link to the foreign object.

Suggestions:
1. Related to point 4 above, it would be useful if any foreign relation is displayed as a link to the foreign object.  Including foreign keys not in raw_id_fields, and including many to many fields.
2. Fields with editable=False (e.g. DateTimeField with auto_now=True) are not displayed in the change form, because they are not editable.  But being editable does not matter if I have only view permission.  So let's show non-editable fields.
3. Use different URLs for viewing and for changing an object.  I will write a separate comment about the benefits of this.

viktor25 commented:
This is about my last suggestion, using different URLs for viewing and for changing an object.

At the moment, the object list page is at /admin/app/somemodel/, and the object details page is at /admin/app/somemodel/123/change/, no matter whether I have change permission or not.

I think we should have two other URLs for read-only mode.  For example /admin/app/somemodel-view/ and /admin/app/somemodel-view/123/change/ (but feel free to suggest alternative URLs).

Pros:
 - Allows looking at an object details in read-only mode even if I have change permission.  Most of the time when I use the admin, I only look for information, without changing anything.  And the read-only page can potentially become a lot better than the change form, because the change form needs to use input widgets, while the read-only page has no such restrictions.
 - If we release Django 2.0 with the current behaviour, it will be more difficult to change it in the future because of backwards compatibility.  It is usually easier to switch from multiple URLs to a single URL than to switch from a single URL to multiple URLs.

Cons:
 - The change list page is at /admin/app/somemodel/ , and there is no obvious nice-looking way to construct a view-only URL from this.  We cannot use /admin/app/somemodel/view/ because it might conflict with the details page for an object with primary key "view".

I'm happy to do the work for this myself it people agree with the idea.  @olivierdalang @timgraham

olivierdalang commented:
@viktor25 Thank you for the review. I don't really mind about URLs, but did you think of the case where the user has read only permissions on the main model, but has add/edit/delete permissions on inlines ? In such case, the distinction between view/change is less clear.

About the regressions you mention, don't they happen as well with regular read-only fields ? If so, I think they are a different (but still very valid) issue and this would deserve another ticket/PR.

I would very much love to see this PR merged for 2.0. Unfortunately, I have almost no time available to work on this in the upcoming weeks, so if you have some time available on this, it would be awesome if you're ready to care of eventual additional change requests.

FingalP commented:
Hello all,

Thanks for all your work on this! I'm working on this with @viktor25, and we're keen to get this into Django 2.0. The freeze date is 18th Sept. We'll try to be ready with our changes by early next week.

Regarding @viktor25's regressions: it's true that the page with view permissions displays the same as if the fields were read-only, so you are correct that should be on a different ticket.

URLs: @olivierdalang, that's a good point. The solution we've thought of is that there should be a view page where you can only view (including the inlines) and, if relevant, a change page, where you can change only the things you have permission to change(e.g. some inlines). This means on /admin/model/ there would be both links to 'Change' and 'View' if you have change permissions and view permissions, or if you have only view permissions but you have change permissions on the inlines. There will only be 'View' only if you have view permissions on the model and don't have change permissions on any inlines.

When the admin has prepopulated_fields, I'm seeing the same error as @slurms. I'll try to make the view page ignore prepopulated_fields.

I think it does make sense to show editable=False fields when you have only view permissions, so I'll work on that too.

FingalP commented:
Regarding uneditable_fields, I agree with @pope1ni's criticisms and I'm also concerned that it sounds a lot like a field with editable=False, which is treated differently by django's admin (e.g. it's not shown if you have change permissions). I could just change the variable used in django/contrib/admin/options.py from uneditable_fields to readonly_fields but it would look strange that I'm not using get_readonly_fields(), which I don't really want to edit.

olivierdalang commented:
@FingalP I removed uneditable_fields altogether (see https://github.com/django/django/pull/6734#issuecomment-310147322) so this shouldn't be a problem anymore.

For me, different view and edit pages is a 👎. I don't really see what problem it solves. The conceptseems a bit flaky (you have a different view depending on permissions on inlines). And your last suggestion about showing uneditable fields anyway seems wrong to me (I think most of the time, uneditable fields are used for technical fields such as timestamps).

I suggest waiting for some more opinions before going this way.

FingalP commented:
Thanks for your comments @olivierdalang. That's good to know about uneditable_fields.

Sorry I wasn't clear about the aim of my comments. The problem which I'm trying to solve is to allow a user with change permissions and view permissions to access the 'view' page, because this can be made loads easier to navigate around and see relevant info, thanks to not having to have input widgets. I get the impression @ryselis has similar concerns for their 'detail' view, allowing for images and charts - and similar implementation, with a separate 'detail' view function. Their tabs idea, for example, would also be possible only if we separate the URLs now, as this would be difficult later otherwise.

You're right we need to make the concept less flaky. Instead of showing different things on the 'view' page depending on the inline permissions, we could just have:
One view page, on which everything is readonly, including inlines (I think fully view-only functionality makes most sense here), but the relevant inlines have links to their change pages,
One change page, on which the model itself can be changed, and any inlines with permission can be modified - i.e. the same as Django's current change page.

This is different to the current behaviour, where the changeform page without change permissions allows you to modify any inlines which you have permission for. I believe the potential reduction in functionality can be fixed by linking to the relevant change pages.

I get your point about editable=False fields, thanks for clarifying. I'll leave them as they are.

FingalP commented:
@olivierdalang I've made the changes I suggested in my last comment, after taking your comments on board. I think I don't have permission to push to this PR (I'm new to GitHub so I'm not sure how this works) but I've made a fork of Django and pushed my current branch here: https://github.com/FingalP/django/tree/view_permission_master. I believe a lot of the rest of the code can be tidied up now that it is mostly contained in the viewlist_view and viewonly_view, though I haven't done much of this yet.

We have less than 2 weeks until the freeze date for Django 2.0, do you know what the next steps are?

aofwer commented:
Hi,

is there any progress on this? Any words from the Django devs? I'd think that this is one of the most-wanted features and it's basically in useable shape.

olivierdalang commented:
Argh some failing tests again. I just posted to ML to see whether they're willing to merge this at some point. If so, I'll try to fix the issues in the upcomming days.

pope1ni commented:
Hi Olivier,

I'd be happy to help try and move this forward again. The consensus on the [ticket](https://code.djangoproject.com/ticket/8936) seems to be that this is accepted, so, assuming there is a PR in a state that is ready to merge in sufficient time before a release, it should be merged. According to the [roadmap](https://code.djangoproject.com/wiki/Version2.1Roadmap) for 2.1 this means that we have until early May 2018 to get this in shape.

It looks like (all of?) the tests are failing with the following error:
```
AttributeError: 'WSGIRequest' object has no attribute 'user'
```
Seems like some issue related to authentication middleware or middleware ordering at a quick guess?

Here is a checklist of things that need to be done:

- [ ] Rebase the branch on top of  master.
- [ ] Fix any obvious failures due to changes in 2.0
- [ ] Perform thorough manual testing of the feature.
- [ ] Check the documentation and release notes are correct.
- [ ] Provide a documented upgrade path to create missing view permissions.

pope1ni commented on docs/ref/contrib/admin/actions.txt:
```diff
@@ -34,6 +34,14 @@ models. For example, here's the user module from Django's built-in
 
+.. note::
+
+    .. versionchanged:: 2.0
```
2.1

pope1ni commented on docs/ref/models/options.txt:
```diff
@@ -326,12 +326,16 @@ Django quotes column and table names behind the scenes.
     the model is created by :djadmin:`migrate` in order to prevent any omitted
     permissions from being created.
 
+    .. versionchanged:: 2.0
```
2.1

pope1ni commented on docs/releases/1.11.txt:
```diff
@@ -87,6 +87,21 @@ Minor features
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
 * :attr:`.ModelAdmin.date_hierarchy` can now reference fields across relations.
+* Added ``view`` permission to the
```
Move to `docs/releases/2.1.txt`

pope1ni commented on docs/releases/2.0.txt:
```diff
@@ -448,6 +454,21 @@ If you wish to keep this restriction in the admin when editing users, set
     admin.site.register(User, MyUserAdmin)
 
+
+Custom admin forms need to take read-only case into account
```
Move to `docs/releases/2.1.txt`

pope1ni commented on docs/releases/2.0.txt:
```diff
@@ -97,6 +97,12 @@ Minor features
 
+* ...
+
+:mod:`django.contrib.admindocs`
```
Undo this unrelated change... (Likely caused by merging master into the branch instead of rebasing.)

olivierdalang commented:
@pope1ni Thanks for the comments. I fixed (I think) the release notes according to your comments.

About the failing tests, it seems they happen because there was some refactoring in `ModelAdmin` (`get_changelist_instance()`) which presupposes the change permission is there, while now user permissions must be checked to see if the user has the change permission.

I don't think testing `get_changelist_instance` with no user makes sense, as `contrib.auth` is required for `contrib.admin`, so I fixed the tests by attaching an user (superuser) to the requests.

Is this correct ? I'm not sure it's OK yet since tests didn't run for my last commit, any idea on how to force tests to run again ?

pope1ni commented on tests/admin_changelist/tests.py:
```diff
@@ -48,6 +48,7 @@ class ChangeListTests(TestCase):
 
     def setUp(self):
         self.factory = RequestFactory()
+        self.superuser = User.objects.create_superuser(username='super', email='a@b.com', password='xxx')
```
Seems sensible based on the [example](https://docs.djangoproject.com/en/1.11/topics/testing/advanced/#example) in the documentation.

The user doesn't need to be a superuser, and I would suggest it is probably better they are not.
Just create a normal staff user:

```python
self.user = User.objects.create_user(username='staff', is_staff=True)
```

olivierdalang replied:
It was just to avoid needing to add the permissions to the user (since it's used so many times)

pope1ni replied:
This also seems strange when I can see a `_create_superuser()` method immediately below `setUp()`.

Are we sure we are doing the correct thing here? Given these tests work in the existing code where there are add, change and delete permissions; what makes adding a view permission require all these changes. We should pinpoint what bit of the new code is causing this and check it.

Further, if we do need to do this, we should probably also be using `_mocked_authenticated_request()`, also below.

carltongibson replied:
> Are we sure we are doing the correct thing here? Given these tests work in the existing code where there are add, change and delete permissions; what makes adding a view permission require all these changes. We should pinpoint what bit of the new code is causing this and check it.

Adding the `request.user` instance to each test is correct. 

Prior to this change, `changelist_view` checks the permissions before calling, `get_changelist_instance()`, which never uses `request.user`, since it's the same for everyone.

https://github.com/django/django/blob/acf9d6404545c58522c3011348d071d6bfca8f69/django/contrib/admin/options.py#L1594-L1598

Now, `get_actions()`, which is called by `get_changelist_instance()`, itself needs to check permissions, so `request.user` becomes requisite. 

A little refactoring (as a separate commit) — to re-use or move the helper methods is probably merited.

olivierdalang commented:
I'd suggest to everybody interested in getting this PR merged to test it (on your projects) and if it works well to make some noise on the ML (maybe in that thread : https://groups.google.com/forum/#!topic/django-developers/WSTQgEo8tHQ seems my voice alone didn't have much effect :) ).

viktor25 commented:
@olivierdalang As far as I can tell, a user with change permission cannot choose to "view" an object - they will always see a form.  Is there a reason for that?

It is important to be able to see the read-only page even if one has change permission, because long-term the read-only page will be more helpful than the form.  For example, the form shows foreign keys as dropdowns, while the read-only page could show them as links to the foreign object.  The read-only page is not constrained by the need to show input widgets, so it is more powerful for visualizing information.

olivierdalang commented:
@viktor25 See [my comment](#issuecomment-326406711) about the rationale for using forms even for the "view" permission.

I'd even add that if we move away from the "admin form" view, I think it won't belong in django.contrib.admin anymore. "View" permission without "change" permission is still meant for staff members. Having different layouts for view permissions would be very confusing for row-based permissions for instance.

I don't know what other think about this ? I don't know how to settle this question, maybe a django core dev could step in ?

And about how read-only fields render (applies to your last suggestion for foreign keys and previous regressions reports), I think some improvement can be done, but as I said, I think this is a different issue (it also applies for regular read-only fields with change permission).

collinanderson commented:
I personally think having separate view and change pages is out of scope and would be much bigger change. If they have edit ability, they should see a form.

"For example, the form shows foreign keys as dropdowns, while the read-only page could show them as links to the foreign object. " - The change page already has links to the foreign object (though maybe that's not obvious enough?)

viktor25 commented:
@olivierdalang

> See my comment about the rationale

From that comment:

> I don't really see what problem it solves.

The problem is that when you only want to view data, input widgets are inefficient (in terms of what kind of information they can show, how much space they take on the page, etc).  A textarea always takes up a minimum amount of space, even if it is empty.  A many-to-many field with filter_horizontal takes a lot of space even if you only have one or two selected foreign objects.  I'm sure that there are more such examples.  So it makes sense to be able to view an object without all those input widgets getting in the way.

> The concept seems a bit flaky (you have a different view depending on permissions on inlines)

In the view page all inlines would be read-only.  What is flaky here?

> And your last suggestion about showing uneditable fields anyway seems wrong to me (I think most of the time, uneditable fields are used for technical fields such as timestamps).

I think it makes more sense to show timestamps because, like you said, the admin is meant for staff members, and timestamps could be very useful for staff members.  I thought that timestamps were not shown in the form just because they cannot be edited.  But if you prefer not to show timestamps (and other uneditable fields), I'm fine with that.

viktor25 commented:
@olivierdalang

> if we move away from the "admin form" view, I think it won't belong in django.contrib.admin anymore

Why not?  Viewing data seems like an important part of what an admin interface does.

> Having different layouts for view permissions would be very confusing for row-based permissions for instance

If by "different layouts" you mean, for example, different field order, then I have been unclear in what I am suggesting.  I am not suggesting changing the layout in the view page.  I imagine that the view page would look exactly like the change page looks if the user has no change permission.

Actually I like your example about row-based permissions.  If we don't have separate view and change pages, then opening two different records of the same model might show one as a form and the other as a read-only page.  But if we have a separate view page, the user can choose to view the two records, and then they will both be shown as read-only pages.  So having a separate view page would be less confusing in the case of row-based permissions.


> And about how read-only fields render (applies to your last suggestion for foreign keys and previous regressions reports), I think some improvement can be done, but as I said, I think this is a different issue

You are right that improving the read-only fields should be part of a different issue.  But you are now laying the "architecture" (for lack of a better word) for how the admin works, so choosing an architecture that makes future improvements possible is part of this issue.

viktor25 commented:
@collinanderson

> would be much bigger change

The view page can be exactly the same as the change page, except that instead of checking for change permissions, the view page pretends that the user has no change permissions for anything.

> The change page already has links to the foreign object

I don't see them in 1.11, maybe they landed later.  But foreign keys were just one example.  Here are some other examples of how read-only widgets can be more useful than form widgets when one doesn't need to change data:
 - JSON fields can be displayed as collapsible trees
 - Empty text fields do not need to take up a whole textarea's worth of space
 - Datetime fields don't need to take two lines each if we don't need to show date/time pickers
 - Many-to-many fields currently either show all objects in the foreign table (which breaks if the foreign table is big) or just show the ids of the selected foreign objects.  A read-only many-to-many widget could show the string representations of just the selected objects.

The point is that there are many, many things that are easier to implement in a read-only fashion.  Implementing them is out of scope, but choosing the right architecture (for lack of a better word) surely is in scope.

pope1ni commented:
@collinanderson 

> I personally think having separate view and change pages is out of scope

Agreed. The concept of read-only fields already exists and the simplest implementation is to leverage this where the user only has the view permission.

@viktor25 

I understand what you are looking for, but it is far beyond the scope of adding view permissions to the admin site.

> It is important to be able to see the read-only page even if one has change permission, because long-term the read-only page will be more helpful than the form.

This would lead to confusion for many users as there are two ways of viewing the same information. It sounds like you are trying to have the admin site - which is there for the purpose of administration and data entry/modification - provide the views that your project/application should provide.

> For example, the form shows foreign keys as dropdowns, while the read-only page could show them as links to the foreign object.

`OneToOneField` and `ForeignKey` show both _add_ and _edit_ icons; `ManyToManyField` only shows an `add` icon. This is the case in Django 1.11.

This PR should ensure that, if there is no _change_ permission, a _view_ icon is shown. Indeed, where the field is read-only, it should also be possible to have a _view_ icon for the related object, or even an _edit_ icon if the user has access to the related model and can change it. (@olivierdalang: can you check this is the case?)

> The read-only page is not constrained by the need to show input widgets, so it is more powerful for visualizing information.

You are free to add your own pages into the admin site or define methods on the `ModelAdmin` to dictate how fields should be rendered. If you feel that the default rendering is not right for certain field types / widgets, open a new ticket for each case (after searching to determine whether the issue has not already been raised.)

Much of your follow-up comments are re-emphasising these same points so I will finish here. You are free to take this up on the mailing list, but it is likely that it would be a separate development if it were agreed. This is purely about adding a new _view_ permission into the existing implementation, not about changing things to provide side-by-side views for different permissions, or to (automatically) improve the look of widgets for those who do not have _change_ permission (as this can be done manually).

mahadi commented:
Is there a chance that this goes into Django 2.1?

olivierdalang commented:
@timgraham This PR is getting very old... Are you willing to merge it ? If not, can you please let us know why ?  If yes, I'll resolve conflicts and squash commits.

timgraham commented:
The scope is just so huge that I haven't had time and motivation to give it a detailed review. Any mistake where the permission isn't appropriately applied could be considered a security issue. If another trusted reviewer does a comprehensive review of this, I'll probably be able to take a second look before the May feature freeze deadline for 2.1.

mahadi commented:
Just again a reminder to not forget about this issue. I think this is a feature which many people would like to have in django. It would be a pity to not have this in the next release as it actually seems to be pretty mature/finished.

auvipy commented:
this will need more work before final review and merge so it would be great to fix them first

mahadi commented:
It would be great to provide a list what needs to be fixed. So working on this topic could be continued and finished in time?

auvipy commented:
Here is a checklist of things that need to be done:

 Rebase and squash the branch on top of master.
 Fix any obvious failures due to changes in 2.0
 Perform thorough manual testing of the feature.
 Check the documentation and release notes are correct.
 Provide a documented upgrade path to create missing view permissions.

this was the list I found

auvipy commented:
I will, please go ahead after that two official django fellow can

mahadi commented:
I will do manual tests as i want to use this in my project any way :) Would love to see this in the next release.

mahadi commented:
Hmm, it seems that the priority of this issue is too low to get one of the needed reviewers working on this... :(

carltongibson commented:
Hi @olivierdalang. As the backlog clears, this has just popped onto my radar. Thank you for your extended effort here. 🙂

> I can do the squash+rebase (including fix for 2.0), but am only willing to do so after I get confirmation from another trusted reviewer that he'll review to PR before next feature freeze.

If you can rebase, and make the other edits, I will review this over the next week. So, assuming there are no more blockers we should be able to make the next deadline. 

(What a monster of a ticket! 👾)

olivierdalang commented:
@carltongibson Thank you very much for looking into that !

I just squashed & rebased on master. There was no tricky conflict, but some new failing tests.

I'll try to fix them. I think it's just because it can now happen that ModelAdmin.get_actions gets called even if there's only the view permission (when displaying the changelist), so it has now to check for user's permissions.

carltongibson commented:
Hi @olivierdalang

Some preliminary comments. More to follow. 

---

First off, can you please tick "Allow edits from maintainers"? 🙂

(I've got some small edits I'd like to push. Let's add incremental changes just now, and squash at the end.) 

---

My initial manual testing is all positive. Installation is fine. Create a user with view-only permissions, browse: all is as expected. So thus far: :+1: 

(More to come here though: I need to check more complicated variations, both for correct-permissions-behaviour and exact button text etc.)

---

>  Provide a documented upgrade path to create missing view permissions.

This is ≈a-non-issue. Running migrate (even for a single app) causes the `post_migrate` signal to fire which runs `create_permissions`, which creates the new `view` permissions (for all content types).

Notes saying: 

* Run `django-admin migrate`, and...
* Update `default_permissions` (if overridden)

... should be sufficient. (@olivierdalang can you think about those for `2.1.txt`?)

pope1ni commented:
@carltongibson - probably also worth noting the possible backward incompatibility if anyone has added their own `view` permission to `default_permissions`?

carltongibson commented:
@pope1ni Yes. Good one. (Although I wonder, if they’d set default_permissions and actually called it `view`, it might just work... need to check)

pope1ni commented:
> it might just work...

Undoubtedly, yes. It is more the case that a permission they may be expecting to serve some specific purpose of their own now provides access to something else.

carltongibson commented on tests/admin_views/admin.py:
```diff
@@ -156,6 +156,10 @@ def has_change_permission(self, request, obj=None):
         """ Only allow changing objects with even id number """
         return request.user.is_staff and (obj is not None) and (obj.id % 2 == 0)
 
+    def has_view_permission(self, request, obj=None):
```
As per comments on `test_change_view` and `test_history_view`, I don't think this is really behaving in the expected way, or really testing the desired behaviour. 

(TODO: pull out the row-level permissions parts of those tests in a separate commit, and then this can be skipped.)

olivierdalang replied:
You mean a separate commit but still in the PR ? (or take it out of the PR altogether and leave a TODO note in the code ?)

carltongibson commented on tests/admin_filters/tests.py:
```diff
@@ -263,7 +263,8 @@ def setUp(self):
         # Users
-        self.alfred = User.objects.create_user('alfred', 'alfred@example.com')
+        # We make requests as alfred, so we set him as superuser (so that he has change permissions on all models)
+        self.alfred = User.objects.create_superuser('alfred', 'alfred@example.com', 'xxx')
```
`alfred` is part of the data set. I think a separate `self.superuser` would be more consistent with the other tests here. 

(Otherwise we see `request.user = self.alfred` in each method and wonder "Who's `alfred`?")

olivierdalang replied:
Ok I'll do that, but this induce some changes in the expected results from the queries in the tests (the commit may be a bit confusing because of that). Let me know if you think it's worth the change.

carltongibson commented on django/contrib/admin/templatetags/admin_modify.py:
```diff
@@ -44,25 +44,38 @@ def submit_row(context):
         ),
-        'show_save_and_continue': not is_popup and context['has_change_permission'] and show_save_and_continue,
-        'show_save': show_save,
+        'show_save_and_continue': not is_popup and can_save and show_save_and_continue,
```
This looks OK on manual testing. 

* I have a user with `add` and `view` permissions on X
* I add an X, using the "Save and continue" button. 
* I'm redirected to the same page to view the object. Only option is "Close without saving".

carltongibson replied:
Except the message displayed isn't quite right in this case: 

```
The X "Great new X" was added successfully. You may edit it again below.
```

It's not true that it may be edited.

pope1ni replied:
Thinking about this again, my point was that "Save and continue" doesn't make sense if you **only** have the *add* permission without *view* or *change* also.

Before you had to have the *change* permission. Now it ought to be *change* or *view*, but has been changed to use `can_save` which allows just having the *add* permission.

This makes no sense. Either we then have to redirect somewhere else other than the change/view view making the behaviour different, i.e. the same as the "Save" button, or we redirect and get a permission denied error.

I haven't tested these scenarios, but just by inspection of the code, it is clear we are changing the expected behaviour here.

pope1ni commented on django/contrib/admin/static/admin/css/forms.css:
```diff
@@ -261,7 +261,7 @@ body.popup .submit-row {
 }
 
-.submit-row input {
+.submit-row input, .submit-row a.button {
```
Why is this added?

Did you intend to remove `line-height` (and other) properties with the same values from later rules?

olivierdalang replied:
You where right, it wasn't really needed... anyways I just refactored the css a little bit

pope1ni commented on django/contrib/admin/templates/admin/submit_line.html:
```diff
@@ -9,5 +9,8 @@
 {% if show_save_and_add_another %}<input type="submit" value="{% trans 'Save and add another' %}" name="_addanother">{% endif %}
 {% if show_save_and_continue %}<input type="submit" value="{% trans 'Save and continue editing' %}" name="_continue">{% endif %}
+{% if show_close %}
+  <a href="{% url opts|admin_urlname:'changelist' %}" class="closelink">{% trans 'Close without saving' %}</a>
```
Will this render as "Close without saving" if we only have view permissions and this cannot save anyway?

olivierdalang replied:
Yes, I think it's better to be clear that there's no saving going on despite clicking a button that's located where the close buttons usually are.

I think it stays quite clear for the user, because that button only appears if there's no editable form on the page anyways.

pope1ni commented on tests/admin_changelist/test_date_hierarchy.py:
```diff
@@ -11,9 +12,14 @@
     factory = RequestFactory()
 
+    def setUp(self):
+        # We make requests as superuser (so that he has change permissions on all models)
```
Please rephrase the comment to avoid personal pronouns.

olivierdalang replied:
About the save and continue: I'll change it so that it only appears when the user can change, and to reword to "Save and view" if the user can only view (good catch !)

pope1ni commented on tests/admin_views/admin.py:
```diff
@@ -156,6 +156,10 @@ def has_change_permission(self, request, obj=None):
 
+    def has_view_permission(self, request, obj=None):
+        """ Only allow viewing objects with id number multiple of 3 """
+        return request.user.is_staff and (obj is not None) and (obj.id % 3 == 0)
```
The parentheses on this line are unnecessary.

olivierdalang replied:
Ok (that was copied from just above, I'm never sure if it's better to keep resemblance with existing code for consistency or do things right)

olivierdalang commented:
Thanks for the detailed review !!

I'll try to find some time next week to deal with them.

olivierdalang commented on tests/admin_views/tests.py:
```diff
@@ -3752,6 +3860,63 @@ def test_simple_inline(self):
         self.assertEqual(Widget.objects.all()[0].name, "Widget 1 Updated")
 
+    def test_simple_inline_permissions(self):
+        "A simple model doesn't perform changes without change permissions for related object"
```
Ok

olivierdalang commented on tests/admin_views/tests.py:
```diff
@@ -3752,6 +3860,63 @@ def test_simple_inline(self):
+        response = self.client.post(collector_url, self.post_data)
+        self.assertEqual(response.status_code, 302)
+        self.assertEqual(Widget.objects.count(), 1)
+        self.assertEqual(Widget.objects.all()[0].name, "Widget 1")
```
Ok

olivierdalang commented on tests/admin_views/tests.py:
```diff
@@ -3623,7 +3721,17 @@ class AdminInlineTests(TestCase):
 
     @classmethod
     def setUpTestData(cls):
-        cls.superuser = User.objects.create_superuser(username='super', password='secret', email='super@example.com')
```
Ok

olivierdalang commented on django/contrib/admin/helpers.py:
```diff
@@ -224,37 +224,49 @@ class InlineAdminFormSet:
+        if has_change_permission:
+            self.viewonly_fields = readonly_fields
+        else:
+            self.viewonly_fields = readonly_fields + flatten_fieldsets(fieldsets)
```
We can't just extend it, because they are different depending on if we are in extra rows (with the add permission, where unless a field is explicitly readonly, you can still change it) or if we are in existing rows (where all fields are readonly if we don't have the change permission).

Still that's not very clear in the code isn't very clear and I'll refactor so it's more readable.

olivierdalang commented on django/contrib/admin/options.py:
```diff
@@ -649,6 +667,11 @@ def get_form(self, request, obj=None, **kwargs):
         readonly_fields = self.get_readonly_fields(request, obj)
         exclude.extend(readonly_fields)
+
+        # If it is a change form and the user has no permission, we exlcude all the fields
```
Ok

olivierdalang commented on django/contrib/admin/options.py:
```diff
@@ -835,6 +858,10 @@ def get_actions(self, request):
             return OrderedDict()
 
+        # Also user has to have change permisson
+        if not self.has_change_permission(request, None):
```
Ok

olivierdalang commented on django/contrib/admin/options.py:
```diff
@@ -1535,12 +1577,15 @@ def _changeform_view(self, request, object_id, form_url, extra_context):
             list(self.get_fieldsets(request, obj)),
             self.get_prepopulated_fields(request, obj),
-            self.get_readonly_fields(request, obj),
+            uneditable_fields,
```
Ok

olivierdalang commented on django/contrib/admin/options.py:
```diff
@@ -1591,7 +1636,7 @@ def changelist_view(self, request, extra_context=None):
         opts = self.model._meta
         app_label = opts.app_label
-        if not self.has_change_permission(request, None):
+        if not self.has_view_permission(request, None) and not self.has_change_permission(request, None):
```
Ok

olivierdalang commented:
I made some progress on this. 

I still need to address the refactoring of the tests cases, and to address https://github.com/django/django/pull/6734#discussion_r184187921 but for the rest, it should be good.

I hope I didn't miss any comment, this whole PR is getting quite confusing :-)

I left everything as separated commits to facilitate review but of course will squash.

Thanks again for your excellent inputs !!

carltongibson commented:
Hi @olivierdalang. Thanks for the effort. 

> ... this whole PR is getting quite confusing :-)

Yes. And there are so many comments now that it's taking (me at least) multiple refreshes to load. 
(Maybe we can move to one final PR once the current batch of comments are addressed.) 

It looks to me as if there are a number of comments not yet addressed. Please take the time to make sure each one is handled. (I don't think we're far off now. 💃🏼)

olivierdalang commented on tests/admin_views/tests.py:
```diff
@@ -3623,7 +3721,17 @@ class AdminInlineTests(TestCase):
+
+        # User who can view Articles
+        cls.permissionuser.user_permissions.add(get_perm(Collector, get_permission_codename('view', Collector._meta)))
+        cls.permissionuser.user_permissions.add(get_perm(Widget, get_permission_codename('view', Widget._meta)))
```
Ok

olivierdalang commented on tests/admin_views/tests.py:
```diff
@@ -3752,6 +3860,63 @@ def test_simple_inline(self):
+        response = self.client.get(collector_url)
+        self.assertContains(response, 'name="widget_set-0-id"')
+
+        # Now resave that inline
```
Indeed this doesn't test much... I can't see why it's there neither. I'll remove it.

I think it's worth testing the 302 response, as it may be that the query is successful, but redirects to a wrong page (e.g. if you don't have view permission and were redirected to change page, you'd get a 403)

olivierdalang commented on tests/admin_views/tests.py:
```diff
@@ -2134,6 +2218,20 @@ def test_overriding_has_module_permission(self):
         self.assertContains(response, sections)
         self.assertNotContains(response, articles)
 
+    def test_overriding_has_module_view_permission(self):
```
Ok

olivierdalang commented on tests/admin_views/tests.py:
```diff
@@ -2096,6 +2169,17 @@ def test_has_module_permission(self):
         self.assertContains(response, 'admin_views')
         self.assertContains(response, 'Articles')
 
+    def test_has_module_view_permission(self):
```
Ok

olivierdalang commented on tests/admin_views/tests.py:
```diff
@@ -1905,7 +1962,9 @@ def test_history_view(self):
         rl2 = RowLevelChangePermissionModel.objects.create(name="even id")
-        for login_user in [self.superuser, self.adduser, self.changeuser, self.deleteuser]:
+
+        logins = [self.superuser, self.viewuser, self.adduser, self.changeuser, self.deleteuser]
```
If OK with you, I suggest not fixing this in this PR (it's already complicated enough :-) )

olivierdalang replied:
Some other tests have the same problem (e.g. test_overriding_has_module_permission that tests for all types of users while the has_module_permission method doesn't even consider the user).

olivierdalang commented on tests/admin_views/tests.py:
```diff
@@ -1729,6 +1757,18 @@ def test_change_view(self):
+        post = self.client.post(article_change_url, change_dict)
+        self.assertEqual(post.status_code, 302)
+        self.assertEqual(Article.objects.get(pk=self.a1.pk).content, '<p>Middle content</p>')
+        self.client.get(reverse('admin:logout'))
```
Ok will add

olivierdalang commented on tests/admin_views/tests.py:
```diff
@@ -1548,6 +1558,24 @@ def test_login(self):
         self.assertEqual(login.status_code, 200)
         self.assertFormError(login, 'form', 'username', ['This field is required.'])
 
+    def test_login_view_user(self):
```
Ok

olivierdalang commented on django/contrib/admin/options.py:
```diff
@@ -2010,10 +2071,19 @@ def _get_form_for_get_fields(self, request, obj=None):
+            # which doesn't have its own individual permissions. The user needs
+            # to have the view permission for the related model in order to
+            # be able to do anything with the intermediate model.
+            return self.has_change_permission(request, obj)
```
Good catch again. Indeed the view permission should be enough to be able to change intermediate models. Will fix this.

olivierdalang commented on tests/admin_views/tests.py:
```diff
@@ -3752,6 +3860,63 @@ def test_simple_inline(self):
+        self.assertEqual(result, True)
+
+        # Without add permission, we can't add a new inline
+        self.post_data['widget_set-0-name'] = "Widget 1"
```
The same post_data is used across other tests that use the same change form. If this is OK for you I'll leave it like this (more consistent with other tests)

olivierdalang commented:
Another round done... I hope I didn't miss anything.

Those may need some additional review (not always sure of the amendments, mostly about test case style) :
https://github.com/django/django/pull/6734#discussion_r183990584
https://github.com/django/django/pull/6734#discussion_r184004926
https://github.com/django/django/pull/6734#discussion_r184005164
https://github.com/django/django/pull/6734#discussion_r184052087
https://github.com/django/django/pull/6734#discussion_r184561811

Hehe I'll make sure to have a bottle of champagne ready for when this gets merged (if it gets merged) !!

carltongibson commented:
Hi @olivierdalang. 

Thanks for the super effort here!

> ...when this gets merged (if it gets merged) !!

It should get merged. It works. And it's been a long-time on the wish list. We just have to cover all the bases. 

Thank you for all your edits. I'd gone through. It looks like you've addressed each point. As such I've squashed and rebased the PR. 

Of the outstanding points: 

* Don't worry about the refactorings, e.g. `post_data` vs `self.post_data` and the row-level permissions tests. We can address those separately. (They're not directly relevant.) 
* You're correct, the `history_view` is read-only, so removing the comment about changes is correct. 
* Given that it affected the query sets elsewhere, I reverted the `alfred` change. We'll keep using `alfred`, and the comment there. 

@pope1ni: I looked at the `add` only permission case. Here from the create form, the instance is created and then you're redirected back to the `app_index`, where you can chose the `Add` button again. @olivierdalang updated the button text so **I think** your query is resolved. Can you confirm?

olivierdalang commented:
@carltongibson @pope1ni 
Thanks again for your involvement in this. Should I squash the commits already ?
Let me know if there's anything else I can do...

carltongibson commented:
Hi @olivierdalang. Sorry for the delay: I was away. 

I squashed and rebased last week, so nothing to do there. 

I've marked this as Ready for checkin on the ticket. Tim will have a look next: it's a big patch, with potential security issues, so it justifies another review. 

Tim, with fresh eyes, will likely make a few more suggestions. I think we've gone over the main scenarios well enough (POSTing data, accessing inlines etc) so hopefully these will just be minor. On that assumption we should be able to get it in. 💃🏼 (But lets await Tim's review.) 

Thanks for your super effort here. This has been a long time on the wish-list!

olivierdalang commented:
Great ! I just redid the commit so that it's properly attributed to @PetrDlouhy (original author) and me.
I see there are now failing tests, it looks like a git related glitch because there's no change in the code.

mahadi commented:
Will this get accepted for the feature freeze (which seems to be today)?

collinanderson commented:
Hi All. I tried this out and it seems to work as expected for me.

One point of feedback, feel free to ignore if you want: I think the button "Close without saving" is a little confusing. It's not like you would ever have any changes that will get discarded, right? Maybe we could say "back to list page" (or have no button at all)?

In any case. This is awesome. Thanks so much!

timgraham commented:
I changed that to say "Close" in my edits that I haven't pushed yet.
>
>

olivierdalang commented:
Yeaaah !!! Finally :D

Thank you all !!!

timgraham commented:
Do you have time to add additional tests as described on the ticket?

On Wed, May 16, 2018 at 6:59 PM olivierdalang <notifications@github.com>
wrote:

> Yeaaah !!! Finally :D
>
> Thank you all !!!
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/django/django/pull/6734#issuecomment-389692026>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAZI3UoXP9AwxjAWfeK46a5dxmTR558-ks5tzK88gaJpZM4IvDfv>
> .
>
